#  1 Welcome to Tidy Data
- For more information, please check the following link authored by Hadley Wickham: [Tidy Data](https://vita.had.co.nz/papers/tidy-data.pdf)

## 1.1 What is tidy data?

When we work with data, the first thing we do before we start making many different visualization or machine learning models would be CLEANING THE DATA.
 
> Tidy Data: Structuring data to facilitate analysis - Hadley Wickham

Real datas are difficult to work with. Also even if we do use the data as it is, a high possibility of making mistakes exist. When we have a difficult dataset to interpret, we want to make analysis easier. 


## 1.2 Cleaning data

The ultimate goal we are trying to get by cleaning up the data is the to get the `Tidy Data`. If you are able to define Tidy data, and be able to transform originl data to it, you may tab on your should by accomplishing your goal of the day.

### Data structure

Before we go into details, we need to ask ouselves this question. How are the data structured?

Data Structure is basically tables made up of rows and columns. Every value is either variable or observation.
- `Variable` contains all values that measure the same underlying attribute across units.
- `Observation` contains all values measured on the same unit across attributes

### Variables & Observations

General rule on determining variables and observations:

1) It is easier to describe functional relationship between variables than between rows

2) It is easier to make comparisons between group of observations than between columns

Now, we can define tidying again as below.

__A STANDARD way of mapping the meaning of a dataset to it’s structure__: How rows, columns, and tables are matched up with observations, variables, and types of data.

### Three functions

The author of the paper that you just saw mainly uses R, so if we’re using Python, we’ll use the following functions: `melt()`, `pivot()`, `pivot_table()`

In order to perform this analysis, the example dataset that we use has to satisfy at least one of the five following conditions:
- Column headers are values, not variable names
- Multiple variables are stored in one column
- Variables are stored in both rows and columns
- Multiple types of observational units are stored in the same table
- A single observational unit is stored in multiple tables



## 1.3 Details on Covid dataset

For this tutorial, we will be using the Covid dataset provided by Johns Hopkins University. The data is extremely big, but for the tutorial purpose, we will strictly work with data on state of Ohio. [Info on the Data](https://github.com/CSSEGISandData/COVID-19/tree/master/csse_covid_19_data/csse_covid_19_time_series)

**Note: The data updates daily, so the result may be slightly different as the time you perform this.**

Let’s first import the packages that we need and then we activate the first five rows of the dataset

In [1]:
import pandas as pd
import numpy as np
import datetime

In [2]:
covid = pd.read_csv("https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_confirmed_US.csv")

In [3]:
covid.head()

UID iso2 iso3  code3    FIPS   Admin2 Province_State Country_Region  \
0  84001001   US  USA    840  1001.0  Autauga        Alabama             US   
1  84001003   US  USA    840  1003.0  Baldwin        Alabama             US   
2  84001005   US  USA    840  1005.0  Barbour        Alabama             US   
3  84001007   US  USA    840  1007.0     Bibb        Alabama             US   
4  84001009   US  USA    840  1009.0   Blount        Alabama             US   

         Lat      Long_  ... 9/24/20  9/25/20  9/26/20  9/27/20  9/28/20  \
0  32.539527 -86.644082  ...    1738     1757     1764     1773     1785   
1  30.727750 -87.722071  ...    5165     5456     5477     5526     5588   
2  31.868263 -85.387129  ...     857      873      882      885      886   
3  32.996421 -87.125115  ...     642      652      654      656      657   
4  33.982109 -86.567906  ...    1594     1608     1611     1617     1618   

   9/29/20  9/30/20  10/1/20  10/2/20  10/3/20  
0     1787     1791     1798     1805     1818  
1     5606     5640     5997     6024     6048  
2      886      896      898      902      921  
3      658      664      672      675      678  
4     1621     1629     1634     1642     1655  

[5 rows x 267 columns]

In [4]:
covid.shape

(3340, 267)

We do have a lot of rows in our data, but we do have many columns.

Since we’re only working with data from state of `Ohio`, we’ll clean what we have.

Let's make a new dataframe and extract all rows where the column `province_state` provides `Ohio`

In [5]:
covid_ohio = covid.loc[covid['Province_State'] == 'Ohio']
print(covid_ohio.shape)
covid_ohio.head()

(90, 267)


UID iso2 iso3  code3     FIPS     Admin2 Province_State  \
2122  84039001   US  USA    840  39001.0      Adams           Ohio   
2123  84039003   US  USA    840  39003.0      Allen           Ohio   
2124  84039005   US  USA    840  39005.0    Ashland           Ohio   
2125  84039007   US  USA    840  39007.0  Ashtabula           Ohio   
2126  84039009   US  USA    840  39009.0     Athens           Ohio   

     Country_Region        Lat      Long_  ... 9/24/20  9/25/20  9/26/20  \
2122             US  38.845411 -83.471896  ...     126      129      131   
2123             US  40.772852 -84.108023  ...    1408     1430     1437   
2124             US  40.847723 -82.272808  ...     236      237      239   
2125             US  41.708603 -80.748302  ...     655      657      659   
2126             US  39.334256 -82.042786  ...     653      669      678   

      9/27/20  9/28/20  9/29/20  9/30/20  10/1/20  10/2/20  10/3/20  
2122      135      136      137      139      144      148      151  
2123     1441     1462     1473     1479     1496     1509     1538  
2124      242      246      248      261      271      280      287  
2125      659      660      662      664      667      670      670  
2126      707      724      740      757      773      793      814  

[5 rows x 267 columns]

In [6]:
covid_ohio = covid.loc[covid['Province_State'] == 'Ohio']
print(covid_ohio.shape)
covid_ohio.head()

(90, 267)


UID iso2 iso3  code3     FIPS     Admin2 Province_State  \
2122  84039001   US  USA    840  39001.0      Adams           Ohio   
2123  84039003   US  USA    840  39003.0      Allen           Ohio   
2124  84039005   US  USA    840  39005.0    Ashland           Ohio   
2125  84039007   US  USA    840  39007.0  Ashtabula           Ohio   
2126  84039009   US  USA    840  39009.0     Athens           Ohio   

     Country_Region        Lat      Long_  ... 9/24/20  9/25/20  9/26/20  \
2122             US  38.845411 -83.471896  ...     126      129      131   
2123             US  40.772852 -84.108023  ...    1408     1430     1437   
2124             US  40.847723 -82.272808  ...     236      237      239   
2125             US  41.708603 -80.748302  ...     655      657      659   
2126             US  39.334256 -82.042786  ...     653      669      678   

      9/27/20  9/28/20  9/29/20  9/30/20  10/1/20  10/2/20  10/3/20  
2122      135      136      137      139      144      148      151  
2123     1441     1462     1473     1479     1496     1509     1538  
2124      242      246      248      261      271      280      287  
2125      659      660      662      664      667      670      670  
2126      707      724      740      757      773      793      814  

[5 rows x 267 columns]

So now we have 90 rows of data, and now let’s check the column names

In [7]:
covid_ohio.columns

Index(['UID', 'iso2', 'iso3', 'code3', 'FIPS', 'Admin2', 'Province_State',
       'Country_Region', 'Lat', 'Long_',
       ...
       '9/24/20', '9/25/20', '9/26/20', '9/27/20', '9/28/20', '9/29/20',
       '9/30/20', '10/1/20', '10/2/20', '10/3/20'],
      dtype='object', length=267)

The data has been collecting daily since `January 22, 2020`. See the last final columns of the dataframe above, you’ll notice columns of dates until today. 

Let’s alter the dataset and eliminate columns we don’t need.


In [8]:
covid_ohio = covid_ohio.iloc[:,5:264]
covid_ohio.drop(covid_ohio.columns[3:6],axis = 1, inplace=True)
covid_ohio

Admin2 Province_State Country_Region  1/22/20  1/23/20  1/24/20  \
2122       Adams           Ohio             US        0        0        0   
2123       Allen           Ohio             US        0        0        0   
2124     Ashland           Ohio             US        0        0        0   
2125   Ashtabula           Ohio             US        0        0        0   
2126      Athens           Ohio             US        0        0        0   
...          ...            ...            ...      ...      ...      ...   
2207  Washington           Ohio             US        0        0        0   
2208       Wayne           Ohio             US        0        0        0   
2209    Williams           Ohio             US        0        0        0   
2210        Wood           Ohio             US        0        0        0   
2211     Wyandot           Ohio             US        0        0        0   

      1/25/20  1/26/20  1/27/20  1/28/20  ...  9/21/20  9/22/20  9/23/20  \
2122        0        0        0        0  ...      121      123      126   
2123        0        0        0        0  ...     1378     1388     1393   
2124        0        0        0        0  ...      222      224      231   
2125        0        0        0        0  ...      651      651      653   
2126        0        0        0        0  ...      616      629      636   
...       ...      ...      ...      ...  ...      ...      ...      ...   
2207        0        0        0        0  ...      250      251      255   
2208        0        0        0        0  ...      905      908      913   
2209        0        0        0        0  ...      199      201      205   
2210        0        0        0        0  ...     1725     1737     1752   
2211        0        0        0        0  ...      201      202      205   

      9/24/20  9/25/20  9/26/20  9/27/20  9/28/20  9/29/20  9/30/20  
2122      126      129      131      135      136      137      139  
2123     1408     1430     1437     1441     1462     1473     1479  
2124      236      237      239      242      246      248      261  
2125      655      657      659      659      660      662      664  
2126      653      669      678      707      724      740      757  
...       ...      ...      ...      ...      ...      ...      ...  
2207      258      260      260      261      262      263      263  
2208      919      927      931      936      940      942      950  
2209      206      207      209      213      214      216      222  
2210     1804     1841     1866     1902     1921     1942     1974  
2211      208      210      212      213      213      215      215  

[90 rows x 256 columns]

In [9]:
# Check if any missing values exist
covid_ohio.isnull().sum()

Admin2            0
Province_State    0
Country_Region    0
1/22/20           0
1/23/20           0
                 ..
9/26/20           0
9/27/20           0
9/28/20           0
9/29/20           0
9/30/20           0
Length: 256, dtype: int64

There are no missing values, so let’s provide an explanation why we will melt and pivot this dataset.

If you see all the dates, this is a prime example of a dataset where `column headers` are `values` and the `variables` are stored in both `rows and columns`.

# 2 Melt, Pivot, Pivot_Table

## 2.1 What is the melt function?

Here’s the page on the melt function: [melt](https://pandas.pydata.org/docs/reference/api/pandas.melt.html)

In order to perform the melt function, we need to send all the columns that have dates as observations and we keep original columns the same and let’s see what happens.

Let's create a variable `covid_ohio_dates` that takes all the columns that have dates.

In [10]:
# Create a dataframe of data we want to work with
covid_ohio_dates = covid_ohio.iloc[:,3:256]
covid_ohio_dates

1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  \
2122        0        0        0        0        0        0        0        0   
2123        0        0        0        0        0        0        0        0   
2124        0        0        0        0        0        0        0        0   
2125        0        0        0        0        0        0        0        0   
2126        0        0        0        0        0        0        0        0   
...       ...      ...      ...      ...      ...      ...      ...      ...   
2207        0        0        0        0        0        0        0        0   
2208        0        0        0        0        0        0        0        0   
2209        0        0        0        0        0        0        0        0   
2210        0        0        0        0        0        0        0        0   
2211        0        0        0        0        0        0        0        0   

      1/30/20  1/31/20  ...  9/21/20  9/22/20  9/23/20  9/24/20  9/25/20  \
2122        0        0  ...      121      123      126      126      129   
2123        0        0  ...     1378     1388     1393     1408     1430   
2124        0        0  ...      222      224      231      236      237   
2125        0        0  ...      651      651      653      655      657   
2126        0        0  ...      616      629      636      653      669   
...       ...      ...  ...      ...      ...      ...      ...      ...   
2207        0        0  ...      250      251      255      258      260   
2208        0        0  ...      905      908      913      919      927   
2209        0        0  ...      199      201      205      206      207   
2210        0        0  ...     1725     1737     1752     1804     1841   
2211        0        0  ...      201      202      205      208      210   

      9/26/20  9/27/20  9/28/20  9/29/20  9/30/20  
2122      131      135      136      137      139  
2123     1437     1441     1462     1473     1479  
2124      239      242      246      248      261  
2125      659      659      660      662      664  
2126      678      707      724      740      757  
...       ...      ...      ...      ...      ...  
2207      260      261      262      263      263  
2208      931      936      940      942      950  
2209      209      213      214      216      222  
2210     1866     1902     1921     1942     1974  
2211      212      213      213      215      215  

[90 rows x 253 columns]

As we melt the data, let `id_vars` maintain the same variables that we ask for, and `value_vars` take all the date columns from a and we make them as observations.

Additionally, the values of all dates are incorporated in a column called `value`.

By doing this, the __number of columns decreases__ from 256 to 5, and the __number of rows increases__ from 90 to 22770.

In [11]:
covid_melt = pd.melt(covid_ohio,
                     id_vars = ['Admin2','Province_State','Country_Region'],
                     value_vars = covid_ohio_dates) 

In [12]:
print(covid_melt.shape)
covid_melt.head()

(22770, 5)


Admin2 Province_State Country_Region variable  value
0      Adams           Ohio             US  1/22/20      0
1      Allen           Ohio             US  1/22/20      0
2    Ashland           Ohio             US  1/22/20      0
3  Ashtabula           Ohio             US  1/22/20      0
4     Athens           Ohio             US  1/22/20      0

## 2.2 What is the pivot function?

Here’s the page on the pivot function: [pivot](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.pivot.html)
    
The goal here is to extract __all rows__ that show the __number of contracted cases__ that happened at a city at a certain data in Ohio.

In [13]:
covid_ohio.head()

Admin2 Province_State Country_Region  1/22/20  1/23/20  1/24/20  \
2122      Adams           Ohio             US        0        0        0   
2123      Allen           Ohio             US        0        0        0   
2124    Ashland           Ohio             US        0        0        0   
2125  Ashtabula           Ohio             US        0        0        0   
2126     Athens           Ohio             US        0        0        0   

      1/25/20  1/26/20  1/27/20  1/28/20  ...  9/21/20  9/22/20  9/23/20  \
2122        0        0        0        0  ...      121      123      126   
2123        0        0        0        0  ...     1378     1388     1393   
2124        0        0        0        0  ...      222      224      231   
2125        0        0        0        0  ...      651      651      653   
2126        0        0        0        0  ...      616      629      636   

      9/24/20  9/25/20  9/26/20  9/27/20  9/28/20  9/29/20  9/30/20  
2122      126      129      131      135      136      137      139  
2123     1408     1430     1437     1441     1462     1473     1479  
2124      236      237      239      242      246      248      261  
2125      655      657      659      659      660      662      664  
2126      653      669      678      707      724      740      757  

[5 rows x 256 columns]

First, melt the dataset for all dates to be now observables

In [14]:
covid_melt = pd.melt(covid_ohio,
                     id_vars = ['Admin2','Province_State','Country_Region'],
                     value_vars = covid_ohio_dates) 

In [15]:
print(covid_melt.shape)
covid_melt.head()

(22770, 5)


Admin2 Province_State Country_Region variable  value
0      Adams           Ohio             US  1/22/20      0
1      Allen           Ohio             US  1/22/20      0
2    Ashland           Ohio             US  1/22/20      0
3  Ashtabula           Ohio             US  1/22/20      0
4     Athens           Ohio             US  1/22/20      0

Now pivot the dataset by:
1. Use the melt output as the dataframe 
2. Pivot the dataset, by making index the list of dates
3. The columns will be every city in Ohio
4. Observations will be the values

In [16]:
covid_pivot = covid_melt.pivot(index='variable', columns='Admin2', values=['value'])
covid_pivot

value                                                               \
Admin2   Adams Allen Ashland Ashtabula Athens Auglaize Belmont Brown Butler   
variable                                                                      
1/22/20      0     0       0         0      0        0       0     0      0   
1/23/20      0     0       0         0      0        0       0     0      0   
1/24/20      0     0       0         0      0        0       0     0      0   
1/25/20      0     0       0         0      0        0       0     0      0   
1/26/20      0     0       0         0      0        0       0     0      0   
...        ...   ...     ...       ...    ...      ...     ...   ...    ...   
9/5/20      95  1206     178       620    419      461     691   234   4702   
9/6/20      96  1222     178       622    424      467     691   235   4769   
9/7/20      96  1234     179       622    434      471     691   236   4799   
9/8/20      96  1244     179       622    437      477     691   239   4837   
9/9/20      98  1254     178       623    447      486     691   242   4906   

                  ...                                                     \
Admin2   Carroll  ... Unassigned Union Van Wert Vinton Warren Washington   
variable          ...                                                      
1/22/20        0  ...          0     0        0      0      0          0   
1/23/20        0  ...          0     0        0      0      0          0   
1/24/20        0  ...          0     0        0      0      0          0   
1/25/20        0  ...          0     0        0      0      0          0   
1/26/20        0  ...          0     0        0      0      0          0   
...          ...  ...        ...   ...      ...    ...    ...        ...   
9/5/20       132  ...          0   393       96     39   2475        233   
9/6/20       132  ...          0   402       97     40   2503        234   
9/7/20       135  ...          0   403      100     40   2528        235   
9/8/20       133  ...          0   412      102     40   2545        236   
9/9/20       135  ...          0   421      102     41   2555        235   

                                       
Admin2   Wayne Williams  Wood Wyandot  
variable                               
1/22/20      0        0     0       0  
1/23/20      0        0     0       0  
1/24/20      0        0     0       0  
1/25/20      0        0     0       0  
1/26/20      0        0     0       0  
...        ...      ...   ...     ...  
9/5/20     810      165  1427     191  
9/6/20     819      167  1438     191  
9/7/20     827      168  1465     192  
9/8/20     831      168  1475     192  
9/9/20     833      170  1484     194  

[253 rows x 90 columns]

__The result clearly shows that the coivd cases increased daily in the state of Ohio.__

## 2.3 What is the pivot_table function?

Here’s the page on the pivot_table function: [pivot_table](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html)

We will use the same objective and obtain the number of cases, but at the end, you will notice the differences in output when the `pivot_table ` function is used.

In [17]:
covid_ohio.head()

Admin2 Province_State Country_Region  1/22/20  1/23/20  1/24/20  \
2122      Adams           Ohio             US        0        0        0   
2123      Allen           Ohio             US        0        0        0   
2124    Ashland           Ohio             US        0        0        0   
2125  Ashtabula           Ohio             US        0        0        0   
2126     Athens           Ohio             US        0        0        0   

      1/25/20  1/26/20  1/27/20  1/28/20  ...  9/21/20  9/22/20  9/23/20  \
2122        0        0        0        0  ...      121      123      126   
2123        0        0        0        0  ...     1378     1388     1393   
2124        0        0        0        0  ...      222      224      231   
2125        0        0        0        0  ...      651      651      653   
2126        0        0        0        0  ...      616      629      636   

      9/24/20  9/25/20  9/26/20  9/27/20  9/28/20  9/29/20  9/30/20  
2122      126      129      131      135      136      137      139  
2123     1408     1430     1437     1441     1462     1473     1479  
2124      236      237      239      242      246      248      261  
2125      655      657      659      659      660      662      664  
2126      653      669      678      707      724      740      757  

[5 rows x 256 columns]

In [18]:
covid_melt = pd.melt(covid_ohio,
                     id_vars = ['Admin2','Province_State','Country_Region'],
                     value_vars = covid_ohio_dates) 

In [19]:
print(covid_melt.shape)
covid_melt.head()

(22770, 5)


Admin2 Province_State Country_Region variable  value
0      Adams           Ohio             US  1/22/20      0
1      Allen           Ohio             US  1/22/20      0
2    Ashland           Ohio             US  1/22/20      0
3  Ashtabula           Ohio             US  1/22/20      0
4     Athens           Ohio             US  1/22/20      0

We’re going to do the same process that we did with the pivot function with a few minor changes:
1. Use the new covid_melt function as our data
2. We are only extracting `observations` as our `values`
3. We will have `two indexes` starting with the data and then the `city name`
4. Our `column` this time will be `Ohio` so we should have an output of one column as a result

In [20]:
covid_pivot_table = pd.pivot_table(covid_melt, values='value',index=['variable','Admin2'], columns=['Province_State'])
covid_pivot_table

Province_State       Ohio
variable Admin2          
1/22/20  Adams          0
         Allen          0
         Ashland        0
         Ashtabula      0
         Athens         0
...                   ...
9/9/20   Washington   235
         Wayne        833
         Williams     170
         Wood        1484
         Wyandot      194

[22770 rows x 1 columns]

As you can see, there are differences between pivot and pivot_table.
- We asked for the same objective, but outputs are different
- `Pivot_table` has an advantage in terms that we can have multiple indexes
- Result: It is fair to say that using `pivot_table`, it is easier to read compared to using `pivot` function

# Assignments

### Q1. General Knowledge

Which of the following is **NOT** an example of a messy dataset?

1. Column headers are values, not variable names
2. Multiple variables are stored in one column
3. Variables are stored in both rows and columns
4. A single observational unit is stored in single tables

Assign your selection to `answer`. (e.g. answer = 5)

In [21]:
# Make your selection below. e.g. answer = 5

# YOUR CODE HERE
answer = 4

In [22]:
assert answer < 5

### Q2. Melt data 

As we can see, `kids` dataset violates rule 1.

Your task is to fix this.
Melt `kids` into a new dataframe: `kids_melted`. Keep default column names.  
Hint: use `parent` column as `id_var`  

In [23]:
kids = pd.DataFrame({
    'parent': ['Frank', 'Fred', 'John', 'Jessica'],
    'daughters': [3, 0, 1, 2],
    'sons': [0, 2, 1, 2]
})

# YOUR CODE HERE
kids_melted = pd.melt(kids,id_vars='parent')

In [24]:
"""Check student's answers"""

assert kids_melted.shape == (8, 3)
assert kids_melted.value.sum() == 11

### Q3. Melt (Challenge Question)

Let's take a look at the covid data below.

We would like melt `4/1/20` to `6/30/20` (Q2 of 2020) on `Admin2`, `Province_State`, `Country_Region`

In [25]:
"""
Steps
    1. Get 'Admin2', 'Province_State', 'Country_Region' & '4/1/20' to '6/30/20' columns and assign it to covid_ohio_Q2
    2. Melt covid_ohio_Q2 to covid_ohio_Q2_melted. Use 'Admin2', 'Province_State', 'Country_Region' as id_vars
    3. Set var_name to Date and value_name to Number_of_Confirmed_Cases
"""

covid = pd.read_csv("https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_confirmed_US.csv") # import data
covid_ohio = covid.loc[covid["Province_State"] == 'Ohio'] # filter down to Ohio

# YOUR CODE HERE
covid_ohio_Q2 = covid_ohio.iloc[:,np.r_[5:8,81:172]]
covid_ohio_Q2_melted = pd.melt(covid_ohio_Q2,
                              id_vars = ['Admin2', 'Province_State', 'Country_Region'],
                              var_name = 'Date',
                              value_name = 'Number_of_Confirmed_Cases')

In [26]:
"""Check student's answers"""

assert covid_ohio_Q2_melted.shape == (8190, 5)  # check shape
assert covid_ohio_Q2_melted.Number_of_Confirmed_Cases.sum() == 2414633  # total confirmed cases

### Q4. Melt (Bonus Question)

What is the total number of confirmed cases in Wood county in Q2?

1. 43890
2. 20896
3. 19559
4. 32578

Assign your selection to `answer`. (e.g. answer = 5)

In [27]:
# Make your selection below. e.g. answer = 5
# Hint use Admin2 == "Wood"

# YOUR CODE HERE
answer = 3

In [28]:
assert answer < 5

### Q5. Pivot

Using the `kids_melted` data above.

Pivot this data using `variable` as index, `parent` as columns and `value` as values.

Assign this to `kids_melted_pivot`

In [39]:
# YOUR CODE HERE
kids_melted_pivot = kids_melted.pivot(index='variable', columns='parent', values='value')

In [40]:
"""Check student's answers"""

assert kids_melted_pivot.shape == (2, 4)  # check shape
assert kids_melted_pivot["Frank"]["daughters"] == 3  # Number of daughters for Frank

### Q6. Pivot (Challenge Question)

Using `covid_ohio_Q2_melted` above, create a pivot using **April Dates** as index, `Admin2` as columns and `Number_of_Confirmed_Cases` as values.  

Assign this to `covid_ohio_april_pivot`

In [31]:
"""
Steps
    1. Make a copy of covid_ohio_Q2_melted as covid_ohio_april
    2. Filter April dates in covid_ohio_april
    3. Pivot covid_ohio_april using Dates as index, Admin2 as columns and Number_of_Confirmed_Cases as values
    4. Assign this to covid_ohio_april_pivot
"""
# YOUR CODE HERE
covid_ohio_april = covid_ohio_Q2_melted
covid_ohio_april = covid_ohio_april.iloc[0:2700]
covid_ohio_april_pivot = covid_ohio_april.pivot(index='Date', 
                                                columns='Admin2', 
                                                values=['Number_of_Confirmed_Cases'])

In [32]:
"""Check student's answers"""

assert covid_ohio_april_pivot.shape == (30, 90)  # check shape
assert covid_ohio_april_pivot.index.nunique() == 30  # check number of dates


### Q7. Pivot (Bonus Question)

What is the total number of confirmed cases in Lucas county on 4/20/20?

1. 0
2. 823
3. 267
4. 973

Assign your selection to `answer`. (e.g. answer = 5)

In [33]:
# Make your selection below. e.g. answer = 5
# Hint use Admin2 == "Wood"

# YOUR CODE HERE
answer = 2

In [34]:
assert answer < 5

### Q8. Pivot Table (Challenge Question)

Using `covid_ohio_Q2_melted` above, create a `Month` column using the `Dates` colum.

Pivot this data using `Admin2` as index, `Month` as columns and sum of `Number_of_Confirmed_Cases` as values.  

Assign this to `covid_ohio_Q2_pivot_table`

Hint make use of `aggfunc` parameter

In [35]:
"""
Steps
    1. Create a Month cloumn based on Date column. 
    2. Pivot covid_ohio_Q2_pivot_table using Admin2 as index, Month as columns and sum of Number_of_Confirmed_Cases as values
"""

covid_ohio_Q2_pivot_table =  covid_ohio_Q2_melted  # copy covid_ohio_Q2_melted
covid_ohio_Q2_pivot_table['Month'] = pd.DatetimeIndex(covid_ohio_Q2_melted['Date']).month_name() # create a Month column

# YOUR CODE HERE
covid_ohio_Q2_pivot_table = pd.pivot_table(covid_ohio_Q2_pivot_table, values='Number_of_Confirmed_Cases',index=['Admin2'], columns=['Month'])

In [36]:
"""Check student's answers"""

assert covid_ohio_Q2_pivot_table.shape == (90, 3) # check shape
assert covid_ohio_Q2_pivot_table.index.nunique() == 90  # check number of dates


### Q9. Pivot Table (Bonus Question)

What is the total number of confirmed cases in Lucas and Wood combined in June?

1. 82987
2. 65734
3. 75245
4. 2987

Assign your selection to `answer`. (e.g. answer = 5)

In [37]:
# Make your selection below. e.g. answer = 5

# YOUR CODE HERE
answer = 3

In [38]:
assert answer < 5

# Conclusion

To summarize this tutorial, we learned the definition and importance of `Tidy data`. By cleaning the data the “Tidy way”, we were easily able to manipulate, model, and observe our data to a simple structure.
Melt(), Pivot(), and Pivot_table() are excellent functions to clean the data. However, there is still much room to improve on. Once everyone starts to collect and organize data, time spent on the analysis of data will be reduced significantly, and solely focus on the analyzing part. From now on, before you do any type of data analysis, cleaning data to a `tidy data` will be the first step. Enjoy analyzing data!!